internal/indexer/transactions_decoder.go

```diff
@@ -0,0 +1,82 @@
+package indexer
+
+import (
+	"encoding/hex"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/Taraxa-project/taraxa-indexer/internal/chain"
+	"github.com/Taraxa-project/taraxa-indexer/internal/common"
+	"github.com/Taraxa-project/taraxa-indexer/internal/contracts"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+)
+
+func splitFunctionIDFromData(data []byte) ([]byte, []byte, error) {
+	if len(data) < 4 {
+		return nil, nil, fmt.Errorf("transaction data is too short")
+	}
+	return data[:4], data[4:], nil
+}
+
+func DecodeTransaction(tx chain.Transaction) (functionSig string, params []string, err error) {
+	if tx.Data == "" {
+		return "", nil, nil
+	}
+	relevantAbi := contracts.ContractABIs[strings.ToLower(tx.To)]
+	if relevantAbi == "" {
+		return "", nil, nil
+	}
+	contractABI, error := abi.JSON(strings.NewReader(relevantAbi))
+	if error != nil {
+		return "", nil, error
+	}
+
+	trimmed := strings.TrimPrefix(tx.Data, "0x")
+	bytes, err := hex.DecodeString(trimmed)
+
+	if err != nil {
+		return "", nil, err
+	}
+
+	funcId, data, err := splitFunctionIDFromData(bytes)
+	if err != nil {
+		return "", nil, err
+	}
+	// Decode the transaction
+	method, err := contractABI.MethodById(funcId)
+
+	functionSig = method.Sig
+
+	if err != nil {
+		return "", nil, err
+	}
+
+	if method == nil {
+		return
+	}
```

Comment:
combine?

Reply:
I'd keep this split as if there is a discrepancy between the ABI version and the deployed contract I wouldn't like to crash the indexing

Reply:
You're assigning error to the `err`, so you don't need to specify it everywhere like this `return "", nil, err`. Also in case of missing method `err == nil`. So code below just doing the same thing 
```
	if err != nil || method == nil {
		return
	}
```